src/lib.rs

```diff
@@ -1,279 +1,132 @@
+        self.0
     }
+}
 
-    fn to_bytes(&self) -> [u8; 96] {
-        let mut result = [0u8; 96];
-        self.write(&mut result[..])
-            .expect("should be able to serialize a FullViewingKey");
-        result
+/// A value that is needed, in addition to a spending key, in order to derive descendant
+/// keys and addresses of that key.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct ChainCode([u8; 32]);
+
+impl ChainCode {
+    /// Constructs a `ChainCode` from the given array.
+    pub fn new(c: [u8; 32]) -> Self {
+        Self(c)
     }
 
-    fn fingerprint(&self) -> FVKFingerprint {
-        let mut h = Blake2b::with_params(32, &[], &[], ZIP32_SAPLING_FVFP_PERSONALIZATION);
-        h.update(&self.to_bytes());
-        let mut fvfp = [0u8; 32];
-        fvfp.copy_from_slice(h.finalize().as_bytes());
-        FVKFingerprint(fvfp)
+    /// Returns the byte representation of the chain code, as required for
+    /// [ZIP 32](https://zips.z.cash/zip-0032) encoding.
+    pub fn as_bytes(&self) -> &[u8; 32] {
+        &self.0
     }
 }
 
-// ZIP 32 structures
-
-/// A Sapling full viewing key fingerprint
-struct FVKFingerprint([u8; 32]);
-
-/// A Sapling full viewing key tag
-#[derive(Clone, Copy, Debug, PartialEq)]
-struct FVKTag([u8; 4]);
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct DiversifierIndex(pub [u8; 11]);
 
-impl FVKFingerprint {
-    fn tag(&self) -> FVKTag {
-        let mut tag = [0u8; 4];
-        tag.copy_from_slice(&self.0[..4]);
-        FVKTag(tag)
+impl Default for DiversifierIndex {
+    fn default() -> Self {
+        DiversifierIndex::new()
     }
 }
 
-impl FVKTag {
-    fn master() -> Self {
-        FVKTag([0u8; 4])
+impl From<u32> for DiversifierIndex {
+    fn from(i: u32) -> Self {
+        u64::from(i).into()
     }
 }
 
-/// A child index for a derived key
-#[derive(Clone, Copy, Debug, PartialEq)]
-pub enum ChildIndex {
-    NonHardened(u32),
-    Hardened(u32), // Hardened(n) == n + (1 << 31) == n' in path notation
-}
-
-impl ChildIndex {
-    pub fn from_index(i: u32) -> Self {
-        match i {
-            n if n >= (1 << 31) => ChildIndex::Hardened(n - (1 << 31)),
-            n => ChildIndex::NonHardened(n),
-        }
+impl From<u64> for DiversifierIndex {
+    fn from(i: u64) -> Self {
+        let mut result = DiversifierIndex([0; 11]);
+        result.0[..8].copy_from_slice(&i.to_le_bytes());
+        result
     }
+}
 
-    fn master() -> Self {
-        ChildIndex::from_index(0)
-    }
+impl TryFrom<DiversifierIndex> for u32 {
+    type Error = std::num::TryFromIntError;
 
-    fn to_index(&self) -> u32 {
-        match self {
-            &ChildIndex::Hardened(i) => i + (1 << 31),
-            &ChildIndex::NonHardened(i) => i,
-        }
+    fn try_from(di: DiversifierIndex) -> Result<u32, Self::Error> {
+        let mut u128_bytes = [0u8; 16];
+        u128_bytes[0..11].copy_from_slice(&di.0[..]);
+        u128::from_le_bytes(u128_bytes).try_into()
     }
 }
```

Comment:
Consider adding `TryFrom<DiversifierIndex> for u64` and `From<DiversifierIndex> for u128`.

Reply:
Opened #7.